lib/collection/src/shards/shard_holder.rs

```diff
@@ -67,28 +70,65 @@ pub enum ShardHashRing {
 }
 
 impl ShardHashRing {
-    /// Create a new single hashring
+    /// Create a new single hashring.
     ///
     /// The hashring is created with a fair distribution of points and `HASH_RING_SHARD_SCALE` scale.
     pub fn single() -> Self {
         Self::Single(HashRing::fair(HASH_RING_SHARD_SCALE))
     }
 
+    /// Create a new resharding hashring, with resharding shard already added into `new` hashring.
+    ///
+    /// The hashring is created with a fair distribution of points and `HASH_RING_SHARD_SCALE` scale.
+    pub fn resharding(shard_id: ShardId) -> Self {
+        let mut ring = Self::Resharding {
+            old: HashRing::fair(HASH_RING_SHARD_SCALE),
+            new: HashRing::fair(HASH_RING_SHARD_SCALE),
+        };
+
+        ring.add_resharding(shard_id);
+
+        ring
+    }
+
     pub fn is_empty(&self) -> bool {
         match self {
             Self::Single(ring) => ring.is_empty(),
             Self::Resharding { old, new } => old.is_empty() && new.is_empty(),
         }
     }
 
+    pub fn is_resharding(&self) -> bool {
+        matches!(self, Self::Resharding { .. })
+    }
+
     pub fn add(&mut self, shard: ShardId) {
         match self {
             Self::Single(ring) => ring.add(shard),
-            // When resharding, update the new ring
-            Self::Resharding { old: _, new } => new.add(shard),
+            Self::Resharding { old, new } => {
+                if new.get(&shard).is_none() {
+                    old.add(shard);
+                    new.add(shard);
+                }
+            }
         }
     }
 
+    pub fn add_resharding(&mut self, shard: ShardId) {
+        if let Self::Single(old) = self {
+            let mut old = mem::replace(old, HashRing::raw());
+            let new = old.clone();
+
+            *self = Self::Resharding { old, new };
+        }
+
+        let Self::Resharding { new, .. } = self else {
+            unreachable!();
+        };
+
+        new.add(shard);
+    }
```

Comment:
- switch from `ShardHashRing::Single` to `ShardHashRing::Resharding`, _if required_
- add shard to the `new` hashring _only_

Reply:
Another point would be: it does not *strictly* starts resharding. When we are loading collection/shard holder, resharding might already be in progress, so in that case we would call `add_resharding` during loading as well.

But I won't fight more over it. I'm cool with either.